trulens_eval/trulens_eval/tru.py

```diff
@@ -214,6 +239,51 @@ def __init__(
                 print(e)
                 self.db = OpaqueWrapper(obj=self.db, e=e)
 
+    @staticmethod
+    def _validate_and_compute_schema_name(name):
+        if not re.match(r"^[A-Za-z0-9_]+$", name):
+            raise ValueError(
+                "`name` must contain only alphanumeric and underscore characters!"
+            )
+        return f"TRULENS_APP__{name.upper()}"
+
+    @staticmethod
+    def _create_snowflake_database_url(
+        snowflake_connection_parameters: Dict[str, str], schema_name: str
+    ) -> str:
+        from snowflake.sqlalchemy import URL
+
+        Tru._create_snowflake_schema_if_not_exists(
+            snowflake_connection_parameters, schema_name
+        )
+        return URL(
+            account=snowflake_connection_parameters["account"],
+            user=snowflake_connection_parameters["user"],
+            password=snowflake_connection_parameters["password"],
+            database=snowflake_connection_parameters["database"],
+            schema=schema_name,
+            warehouse=snowflake_connection_parameters.get("warehouse", None),
+            role=snowflake_connection_parameters.get("role", None),
+        )
+
+    @staticmethod
+    def _create_snowflake_schema_if_not_exists(
+        snowflake_connection_parameters: Dict[str, str], schema_name: str
+    ):
+        from snowflake.core import CreateMode
+        from snowflake.core import Root
+        from snowflake.core.schema import Schema
+        from snowflake.snowpark import Session
```

Comment:
How do we evaluate between using snowpark vs python connector? https://docs.snowflake.com/en/developer-guide/python-connector/python-connector-connect

Reply:
Sorry, what do you mean by "evalulate"?

Reply:
AFAICT both are needed cuz python connector, while it holds active session information and supports setting params on existing session, does not expose (at least I wasn't able to find a straightforward api) Session obj directly.

Reply:
Just talked with sumit, seems like anything to create a stored proc or udf requires snowpark session. So this is good.